EIPS/eip-00.md

```diff
@@ -0,0 +1,831 @@
+
+Returns the return value of [the `sender` execution frame](#sender-execution-frame).
+
+If provided with `paymasterData` and `deployerData` also executes the corresponding frame.
+
+If any of the frames reverts the call returns the revert data of each reverted frame.
+
+#### `eth_estimateGasAccountAbstraction`
+
+Accepts Transaction Type `AA_TX_TYPE` with fields `validationGasLimit`, `paymasterGasLimit`, `callGasLimit` optional.
+
+Optionally accepts the State Override Set to allow users to modify the state during the gas estimation.
+This field as well as its behavior is equivalent to the ones defined for `eth_call` RPC method.
+
+Returns `{validationGasLimit, paymasterGasLimit, callGasLimit, builderFee}` object.
+
+Note that the `deployerData` and `paymasterData` fields are required for a consistent result.
+
+As mentioned earlier, the `sender` and `paymaster` contracts should not revert on the validation failure
+and should return a value different from `MAGIC_VALUE_SENDER` or `MAGIC_VALUE_PAYMASTER` accordingly
+in order to enable gas estimation.
+
+One acceptable way to achieve this behavior for Smart Contract Accounts is to compare the `signature` parameter to
+a predetermined "dummy signature" and to return without reverting in case the values match.
+This will not result in transaction being authorized as long as returned value does not include `MAGIC_VALUE_SENDER`.
+
+## Rationale
+
+### Using Solidity method selectors in a Core EIP
+
+The contracts that have a role in this Account Abstraction proposal, such as `sender` or `paymaster`,
+MUST know which code to execute and understand the calldata provided to them in order to validate the transaction.
+
+We argue that the most straightforward implementation is to rely on Solidity 4-byte method selectors as it is an
+established de-facto standard.
+
+### Accepting `AA_TX_TYPE` transactions from EOAs
+
+While it may seem like allowing EOAs to initiate `AA_TX_TYPE` transactions contradicts the purpose of Account Abstraction, we argue that this
+may actually be important for the adoption of Smart Contract Accounts.
+
+It will enable all existing EOAs to benefit from the improved UX features like gas abstraction and validity ranges.
+
+In the future, this can be used to pay gas for transactions that add code to the EOA addresses,
+once Ethereum implements changes like the ones proposed in
+[EIP-5003: Insert Code into EOAs with AUTHUSURP](./eip-5003),
+[EIP-6913: SETCODE instruction](./eip-6913) and
+[EIP-7377: Migration Transaction](./eip-7377).
+
+## Backwards Compatibility
+
+This EIP preserves most of the design elements established by the ERC-4337. This allows the same client code and smart
+contracts to be used in both systems with minimal to no modifications, while providing significant UX improvements.
+
+Existing contracts are not significantly affected by the change.
+The assumption that `tx.origin` is guaranteed to be an EOA is no longer valid.
```

Comment:
Thanks.  I'm familiar with that report, but I think we need to consider the pros and the cons here.  If we keep the existing `tx.origin` behavior, we're effectively enshrining EOA and negating the value of native AA.  `tx.origin` checks have been the biggest obstacle to wide AA adoption (along with lack of EIP-1271 support, which is [improving rapidly](https://eip1271.io/)).  @juniset may want to weigh in on this one, being one of the people who had to deal with this problem for a few years.

IMHO the community has been given sufficient time to phase tx.origin out.  There has been discussions about deprecating it [since 2016](https://github.com/ethereum/solidity/issues/683), and any tutorial from the past few years warns against using it.

tx.origin also gives a false sense of security and doesn't actually prevent attacks.  In the flash loan example, a block builder may include three transactions bundled together, emulating the behavior of the flash loan from an EOA. It would fund the EOA with a large amount, perform the attack, and move the funds back to safety.

If this change is controversial, we could move it to a separate RIP, letting each L2 decide for itself whether it wants to include it in addition to the native AA one.  Personally I think that not including it would reduce usability, but that's up to each L2 to decide (and up to @juniset to decide whether Argent will support an L2 that doesn't 😉).